File: LEETCODE/maximum_score_from_removingString_1717.py
class Solution:
    def maximumGain(self, s: str, x: int, y: int) -> int:
        def solve(s, x, y):
            if len(s) <= 0:
                return 0
            pos1 = s.find("ab");
            score1 = 0;
            if (pos1 != -1):
                score1 = solve(s.strip("ab"), x, y) + x
            pos2 = s.find("ba")
            score2 = 0;
            if (pos2 != -1):
                score2 = solve(s.strip("ba"), x, y) + y
            return max(score1, score2)
        return solve(s, x, y)

File: LEETCODE/maximum_score_from_removingString_1717.cpp
#include<iostream>
#include<unordered_map>
#include <stack>
#include <vector>
using namespace std;

class Solution {
public:// ab
    pair<string, int> counter(string &s, int val, string sub){
        string st;
        int score = 0;
        for(auto ele:s){
            if (st.empty()){
                st.push_back(ele);
                continue;
            }
            if (ele == sub[1] && st.back() == sub[0]){
                score += val;
                st.pop_back();
            }else{
                st.push_back(ele);
            }
        }
        return make_pair(st, score);
    }
    int solve(string s, int x, int y, unordered_map<string, int>& dp){
        int score = 0;
        if (x>y){
            pair<string, int> temp;
            temp = counter(s, x, "ab");
            score = temp.second + counter(temp.first, y, "ba").second;
        }else{
            pair<string, int> temp;
            temp = counter(s, x, "ba");
            score = temp.second + counter(temp.first, y, "ab").second;
        }
        return score;
    }

    int maximumGain(string s, int x, int y) {
        unordered_map<string, int> dp;
        return solve(s, x, y, dp);
    }
};